expression/partition_pruner.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/disjointset"
+)
+
+type hashPartitionPruner struct {
+	unionSet    *disjointset.IntSet // unionSet stores the relations like col_i = col_j
+	constantMap []*Constant
+	conditions  []Expression
+	colMapper   map[int64]int
+	numColumn   int
+	ctx         sessionctx.Context
+}
+
+func (p *hashPartitionPruner) getColID(col *Column) int {
+	return p.colMapper[col.UniqueID]
+}
+
+func (p *hashPartitionPruner) insertCol(col *Column) {
+	_, ok := p.colMapper[col.UniqueID]
+	if !ok {
+		p.numColumn += 1
+		p.colMapper[col.UniqueID] = len(p.colMapper)
+	}
+}
+
+func (p *hashPartitionPruner) reduceColumnEQ() bool {
+	p.unionSet = disjointset.NewIntSet(p.numColumn)
+	for i := range p.conditions {
+		if fun, ok := p.conditions[i].(*ScalarFunction); ok && fun.FuncName.L == ast.EQ {
+			lCol, lOk := fun.GetArgs()[0].(*Column)
+			rCol, rOk := fun.GetArgs()[1].(*Column)
+			if lOk && rOk {
+				lID := p.getColID(lCol)
+				rID := p.getColID(rCol)
+				p.unionSet.Union(lID, rID)
+			}
+		}
+	}
+	for i := 0; i < p.numColumn; i++ {
```

Comment:
Loop once seems not enough for the propagation process,  with newly added column equality `unionSet` ?